### ISSUE TITLE
[Containers] Connecting VS Code Remote to container runtimes other than Docker (containerd, CRI-O)

### ISSUE THREAD

**Issue by MurzNN** [microsoft/vscode-remote-release#6014]
Kubernetes [deprecated Docker as a container runtime after v1.20](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) at the end of 2020 year with removing in v1.22 in late 2021.

So now there is more and more servers, that uses other container runtimes (containerd, CRI-O) instead of Docker.

But [Visual Studio Code Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension doesn't see CRI containers, that using runtime other than Docker, because instead of list of containers I see the placeholder:
```
Get started with containers by installing Docker or by visiting the help view.
```

So, is it possible to add support for other container runtimes to `VS Code Remote Containers` extension?

**Comment by MurzNN** [comment]
Now we have [nerdctl](https://github.com/containerd/nerdctl) as a Docker-compatible CLI for containerd, maybe we can reuse it instead of docker backend to manage container operations for VS Code Remote Containers extension?

**Comment by MurzNN** [comment]
Removing Dockershim and Docker support from Kubernetes[ is scheduled for Kubernetes v1.24](https://kubernetes.io/blog/2020/12/02/dockershim-faq/#when-will-dockershim-be-removed).

**Comment by chrmarti** [comment]
You can use a Docker compatible CLI by changing this setting:
<img width="526" alt="image" src="https://user-images.githubusercontent.com/9205389/145612903-e25ac579-790d-44d2-b792-a9f437be680c.png">

When I tried nerdctl with lima a while ago, it looked like it was not fully compatible with the Docker CLI yet.

**Comment by Chuxel** [comment]
We should make this more specific about `nerdctl`. 

Remote - Containers does not touch engines or daemons directly but rather interacts with CLIs. It makes no assumptions about how the engine functions .

There's no "containerd, CRI-O" CLI per-se.  For example, Docker itself uses containerd on Linux. Each CLI that needs to be supported will require separate work unless it supports everything docker does. The nerdctl CLI is part of the containerd project but currently has gaps. 

Updating the subject to be more specific.

**Comment by MurzNN** [comment]
`nerdctl` is only one option, most easier to integrate (because of Docker compatibility), but we have also [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) that must give compatibility with any CRI-compatible container engine!

So maybe split the issue to 2 parts - nerdctl and crictl?

**Comment by Chuxel** [comment]
We should track `crictl` as a separate issue. Each CLI will be a separate effort. Raised #6075

**Comment by Chuxel** [comment]
FYI - For those not aware, the dev container CLI is now open source and is what the extension uses to interoperate with container engines.  https://github.com/devcontainers/cli

**Comment by Anutrix** [comment]
Any updates on this?

**Comment by kidbrax** [comment]
> FYI - For those not aware, the dev container CLI is now open source and is what the extension uses to interoperate with container engines if anyone has interest in contributing. https://github.com/devcontainers/cli

@Chuxel I checked the devcontainers/cli repo and [searched for](https://github.com/search?q=repo%3Adevcontainers%2Fcli+APIVersion&type=code) `APIVersion` and got no results. This leads me to believe the change we need is not in that repo, since the command VSCode runs when spinning up a devcontainer (when trying finch) is `finch version --format {{.Server.APIVersion}}`.

Is the code for this extension not open? This repo is just for feedback, correct?

**Comment by Chuxel** [comment]
> > FYI - For those not aware, the dev container CLI is now open source and is what the extension uses to interoperate with container engines if anyone has interest in contributing. https://github.com/devcontainers/cli
> 
> [@Chuxel](https://github.com/Chuxel) I checked the devcontainers/cli repo and [searched for](https://github.com/search?q=repo%3Adevcontainers%2Fcli+APIVersion&type=code) `APIVersion` and got no results. This leads me to believe the change we need is not in that repo, since the command VSCode runs when spinning up a devcontainer (when trying finch) is `finch version --format {{.Server.APIVersion}}`.
> 
> Is the code for this extension not open? This repo is just for feedback, correct?

Yeah, I'd recommend opening an issue in the CLI repository to discuss the specific scenario you are looking into. 

@chrmarti is the version check in the CLI or extension? Generally, seems like these kinds of checks should end up in the CLI (if not the broader spec) so they can be contributed, but I'm not sure if this is a proposed feature.

**Comment by chrmarti** [comment]
The version check is in the extension. Making that more tolerant now.

**Comment by jsamuel1** [comment]
nerdctl also doesn't support --sig-proxy=false
This appears to be a fixed parameter in devcontainer when using single-container devcontainer configurations.
https://github.com/containerd/nerdctl/blob/main/docs/command-reference.md#whale-nerdctl-attach

(Offending Line in devcontainer code:  https://github.com/devcontainers/cli/blob/f7d4c853bf8c284d784173f3e915a34d961b0b55/src/spec-node/singleContainer.ts#L396  )

**Comment by chrmarti** [comment]
I'm hesitant to make details like `--sig-proxy=false` conditional on the detected CLI as that might quickly become a long list of cases to distinguish given the growing list of Docker-like CLIs.

The more tolerant version check is available in Dev Containers extension 0.330.0-pre-release.

**Comment by Bigous** [comment]
> I'm hesitant to make details like `--sig-proxy=false` conditional on the detected CLI as that might quickly become a long list of cases to distinguish given the growing list of Docker-like CLIs.

Hi @chrmarti ,

In this case, for docker, you are tuning off the feature... and this feature doesn't even exists in nerdctl... you can safely be "non conditional" just detecting you are using nerdctl (which it already do) and not sendind the argument.

Could that be a solution?

**Comment by CodeChanning** [comment]
A --sig-proxy change was merged to nerdctl yesterday that should supports the flag in run
https://github.com/containerd/nerdctl/pull/3043

**Comment by midnight-wonderer** [comment]
I tried substituting the `docker` command with `nerdctl` using the `--docker-path` option.  
The issue I encountered was with the [`additional_contexts` support detection logic](https://github.com/devcontainers/cli/blob/1769aed60fe19a03ef75b4115c5ca4870cb30e43/src/spec-node/dockerCompose.ts#L191), and possibly also with `parseVersion`.

The CLI always determined that `nerdctl compose` was unsupported, which caused the build to fail.  
> ERROR [internal] load metadata for docker.io/library/dev_container_feature_content_temp:latest

For reference, my `nerdctl` is integrated with version `0.19`:
~~~
 buildctl:
  Version:	0.19.0
~~~

I wish the `docker compose` module of the CLI to be refactored into a class, so we can inject the differences as functions into the constructor (or extract the differences to the constructor).

**Comment by cmullark** [comment]
I tried this today with version 0.409.0 of this extension with finch, which uses nerdctl, instead of docker and I am able to spin up a devcontainer!

Not sure where the rough edges are yet but this is now possible today.

**Comment by subesokun** [comment]
Finch's devcontainer Mac integration worked for me: [https://runfinch.com/docs/integrations/mac/devcontainers](https://runfinch.com/docs/integrations/mac/devcontainers). Consider tweaking the default config (cups, memory, ...): [https://runfinch.com/docs/configuration-reference/](https://runfinch.com/docs/configuration-reference/).
